16-encoding-categorical-data.Rmd

```diff
@@ -1,8 +1,356 @@
+  mutate(Hash = strtoi(substr(Hash, 26, 32), base = 16L),  
+         ## now take the modulo
+         Hash = Hash %% 16) %>%
+  select(Neighborhood, Hash)
+```
+
+Now instead of the `r n_distinct(ames_train$Neighborhood)` neighborhoods in our original data or an incredibly huge number of the original hashes, we have sixteen hash values. This method is very fast and memory efficient, and it can be a good strategy when there are a large number of possible categories.
+
+We can implement feature hashing using a recipe step from the `r pkg(textrecipes)` package.
+
+:::rmdnote
+Feature hashing is useful for text data as well as high cardinality categorical data. See Section 6.7 of @Hvitfeldt2021 for a case study demonstration with text predictors.
+:::
+
+```{r}
+library(textrecipes)
+ames_hash <- 
+  recipe(Sale_Price ~ Neighborhood + Gr_Liv_Area + Year_Built + Bldg_Type + 
+           Latitude + Longitude, data = ames_train) %>%
+  step_log(Gr_Liv_Area, base = 10) %>% 
+  step_dummy_hash(Neighborhood, signed = FALSE, num_terms = 16L) %>%
+  step_dummy(all_nominal_predictors()) %>% 
+  step_interact( ~ Gr_Liv_Area:starts_with("Bldg_Type_") ) %>% 
+  step_ns(Latitude, Longitude, deg_free = 20)
+
+ames_hash
+```
+
+Feature hashing is fast and efficient but has a few downsides. For example, different category values often map to the same hash value. This is called a _collision_ or _aliasing_. How often did this happen with our neighborhoods in Ames?
+
+```{r echo=FALSE}
+hash_table <-
+  prep(ames_hash) %>% 
+  bake(new_data = NULL, starts_with("Neighborhood_")) %>%
+  bind_cols(ames_train %>% select(Neighborhood)) %>%
+  distinct() %>%
+  select(-Neighborhood) %>%
+  map_dbl(sum) %>%
+  enframe() 
+
+hash_table %>%
+  kable(col.names = c("Hash", "Number of neighborhoods")) %>%
+  kable_styling(full_width = FALSE)
+```
+
+The number of neighborhoods mapped to each hash value varies between `r min(hash_table$value)` and `r max(hash_table$value)`. All of the hash values greater than one are examples of hash collisions.
+
+- Feature hashing is not directly interpretable because hash functions cannot be reversed. We can't determine what the input category levels were from the output or if a collision occurred.
+
+- The number of hash values is a _tuning parameter_ of this preprocessing technique, and you should try several values to determine what is best for your particular modeling approach. A lower number of hash values results in more collisions, but a high number may not be an improvement over your original high cardinality variable.
+
+- Feature hashing can handle new category levels at prediction time, since it does not rely on pre-determined dummy variables.
+
+- You can reduce hash collisions with a _signed_ hash by using `signed = TRUE`.
+
+:::rmdwarn
+It is likely that some hash columns will contain all zeros, as we see in this example. We recommend a zero-variance filter via `step_zv()` to filter such columns out.
+:::
+
+## Entity embeddings
```

Comment:
I'm torn about this section. I think that the technique has merit when there are one or more really high cardinality predictors. Look at this example (and the others that we've used) I think that readers might walk away thinking that it isn't effective. 

Maybe, instead of this chapter, give a short overview of tolls (including this) for transforming factors to numbers. It could include some parts about ordinal predictors too.

Reply:
I moved in the section on ordered factors from the recipes chapter, and made a shorter "overview" last section.